src/mumble/GlobalShortcut_win.cpp

```diff
@@ -130,576 +52,356 @@ GlobalShortcutWin::~GlobalShortcutWin() {
+		!= HIDP_STATUS_SUCCESS) {
+		return;
 	}
 
-	ql << keyid;
-	ql << QVariant(QUuid(GUID_SysKeyboard));
-	bool suppress = gsw->handleButton(ql, down);
+	device.buttons.assign(device.buttons.size(), false);
+	for (ULONG i = 0; i < nUsages; ++i) {
+		device.buttons[usages[i] - device.usageRange.first] = true;
+	}
 
-	return suppress;
+	for (size_t i = 0; i < device.buttons.size(); ++i) {
+		const QList< QVariant > ql{ QVariant::fromValue(DeviceTypes::HID), QString::fromStdString(device.name),
+									QString::fromStdString(device.prefix), i };
+		handleButton(ql, device.buttons[i]);
+	}
 }
 
-bool GlobalShortcutWin::handleMouseMessage(unsigned int btn, bool down) {
-	GlobalShortcutWin *gsw = static_cast< GlobalShortcutWin * >(engine);
-
-	bool suppress = false;
+void GlobalShortcutWin::on_keyboardMessage(const uint16_t flags, uint16_t scanCode, const uint16_t virtualKey) {
+	const auto up = flags & RI_KEY_BREAK;
+	const auto e0 = flags & RI_KEY_E0;
+	const auto e1 = flags & RI_KEY_E1;
 
-	if (btn > 0) {
-		QList< QVariant > ql;
-		ql << static_cast< unsigned int >((btn << 8) | 0x4);
-		ql << QVariant(QUuid(GUID_SysMouse));
+	if (virtualKey == VK_NUMLOCK) {
+		// Correct PAUSE/BREAK and NUM LOCK silliness, and set the extended bit.
+		scanCode = (MapVirtualKey(virtualKey, MAPVK_VK_TO_VSC) | 0x100);
+	}
 
-		// Do not suppress LBUTTONUP though (so suppression can be deactivated via mouse).
-		bool wantsuppress = gsw->handleButton(ql, down);
-		suppress          = wantsuppress && (btn != 3);
+	// E0 and E1 are escape sequences used for certain special keys, such as PRINT and PAUSE/BREAK.
+	// See https://www.win.tue.nl/~aeb/linux/kbd/scancodes-1.html.
+	if (e1) {
+		// For escaped sequences, turn the virtual key into the correct scan code using MapVirtualKey().
+		// MapVirtualKey() is unable to map VK_PAUSE (this is a known bug), hence we map that by hand.
+		scanCode = virtualKey != VK_PAUSE ? MapVirtualKey(virtualKey, MAPVK_VK_TO_VSC) : 0x45;
 	}
 
-	return suppress;
+	const QList< QVariant > ql{ QVariant::fromValue(DeviceTypes::KEYBOARD), scanCode, e0 };
+	handleButton(ql, !up);
 }
 
-LRESULT CALLBACK GlobalShortcutWin::HookKeyboard(int nCode, WPARAM wParam, LPARAM lParam) {
-	GlobalShortcutWin *gsw = static_cast< GlobalShortcutWin * >(engine);
-	KBDLLHOOKSTRUCT *key   = reinterpret_cast< KBDLLHOOKSTRUCT * >(lParam);
-
-#ifndef QT_NO_DEBUG
-	static int safety = 0;
+void GlobalShortcutWin::on_mouseMessage(const uint16_t flags, const uint16_t data) {
+	uint8_t button;
+	bool down;
 
-	if ((++safety < 100) && (nCode >= 0)) {
-#else
-	if (nCode >= 0) {
-#endif
-		DWORD scancode = key->scanCode;
-		DWORD vkcode   = key->vkCode;
-		bool extended  = !!(key->flags & LLKHF_EXTENDED);
-		bool up        = !!(key->flags & LLKHF_UP);
-		bool suppress  = handleKeyboardMessage(scancode, vkcode, extended, !up);
-		if (suppress) {
-			return 1;
-		}
+	switch (flags) {
+		case RI_MOUSE_BUTTON_1_DOWN:
+			// Left button pressed
+			button = 1;
```

Comment:
Where are these numbers coming from? Do they have an inherent meaning other than that this is mouse button number x? (e.g. are these supposed to be keycodes of Qt or something like this?)

If these are our internal codes only, then I would suggested creating named constants like
```cpp
constexpr uint8_t MOUSE_LEFT_BUTTON = 1;
```
and so on.

Reply:
The number identifies the button and is also what is shown to the user.

Numbers are consistent with the ones in the defines. For example: `RI_MOUSE_BUTTON_1_DOWN` -> 1.

Reply:
Still I think it'd be better to have named constants instead of raw literals. You never know when you might have to reference these values somewhere else again